File: frontend/src/resolver/spotifyResolver.js
import { Resolver } from './resolver.js';
import { getValidAccessToken } from '../auth/spotifyAuth.js';

// ── Keyword sets for flag derivation ────────────────────────────────────────

const LIVE_KEYWORDS = [
  'live', 'in concert', 'live at', 'live from', 'live session',
  'live version', 'live performance', 'acoustic live',
];

const REMIX_KEYWORDS = [
  'remix', 'remixed', 'sped up', 'slowed', 'nightcore', 'flip',
  'bootleg', 'edit', 'vip', 'extended mix', 'club mix', 'radio edit',
  'mashup', 'version', 're-edit',
];

/**
 * Returns true if the track name (lowercased) contains any of the keywords.
 * @param {string} name
 * @param {string[]} keywords
 */
function containsKeyword(name, keywords) {
  const lower = name.toLowerCase();
  return keywords.some((kw) => lower.includes(kw));
}

// ── Normalise a raw Spotify track object into a Candidate ───────────────────

/**
 * @param {object} track  - Raw Spotify track object from the search response
 * @param {'field-qualified'|'plain-combined'|'title-only'} queryRung
 * @returns {import('./resolver.js').Candidate}
 */
function normalise(track, queryRung) {
  return {
    id:          track.id,
    title:       track.name,
    artist:      track.artists?.[0]?.name ?? null,
    artists:     track.artists?.map((a) => a.name).join(', ') ?? null,
    album:       track.album?.name ?? null,
    imageUrl:    track.album?.images?.[2]?.url ?? track.album?.images?.[1]?.url ?? track.album?.images?.[0]?.url ?? null,
    releaseYear: track.album?.release_date ? track.album.release_date.split('-')[0] : null,
    popularity:  track.popularity ?? 0,
    durationMs:  track.duration_ms ?? 0,
    isLive:      containsKeyword(track.name, LIVE_KEYWORDS),
    isRemix:     containsKeyword(track.name, REMIX_KEYWORDS),
    queryRung,
  };
}

// ── SpotifyResolver ──────────────────────────────────────────────────────────

export class SpotifyResolver extends Resolver {
  /**
   * @param {number} [limit=10]  - Max candidates to return per search
   */
  constructor(limit = 10) {
    super();
    this.limit = limit;
  }

  /**
   * Calls Spotify's search endpoint with a 3-rung fallback ladder:
   *   1. field-qualified  — track:"{title}" artist:"{artist}"
   *   2. plain-combined   — "{title}" "{artist}"
   *   3. title-only       — "{title}"
   *
   * Stops at the first rung that returns ≥ 1 result.
   * Logs which rung succeeded (or that all rungs failed) to the console.
   *
   * @param {string}      title
   * @param {string|null} [artist]
   * @returns {Promise<import('./resolver.js').Candidate[]>}
   */
  async search(title, artist = null) {
    const token = await getValidAccessToken();

    const rungs = buildRungs(title, artist);

    for (const { label, query } of rungs) {
      const tracks = await searchSpotify(query, this.limit, token);

      if (tracks.length > 0) {
        console.info(`[SpotifyResolver] "${title}"${artist ? ` / "${artist}"` : ''} → rung: ${label} (${tracks.length} results)`);
        return tracks.map((t) => normalise(t, label));
      }

      console.info(`[SpotifyResolver] "${title}" rung "${label}" → 0 results, trying next…`);
    }

    console.warn(`[SpotifyResolver] "${title}" — all rungs exhausted, no results.`);
    return [];
  }
}

// ── Query ladder builder ─────────────────────────────────────────────────────

/**
 * Builds the ordered list of (label, query) pairs for the ladder.
 * Rungs that don't apply (e.g. no artist → skip field-qualified & plain-combined)
 * are omitted so we never fire pointless requests.
 *
 * @param {string}      title
 * @param {string|null} artist
 * @returns {Array<{label: string, query: string}>}
 */
function buildRungs(title, artist) {
  const rungs = [];

  if (artist) {
    rungs.push({
      label: 'field-qualified',
      query: `track:"${title}" artist:"${artist}"`,
    });
    rungs.push({
      label: 'plain-combined',
      query: `"${title}" "${artist}"`,
    });
  }

  rungs.push({
    label: 'title-only',
    query: `"${title}"`,
  });

  return rungs;
}

// ── Raw Spotify search call ──────────────────────────────────────────────────

/**
 * @param {string} query
 * @param {number} limit
 * @param {string} token
 * @returns {Promise<object[]>}  raw Spotify track objects
 */
async function searchSpotify(query, limit, token) {
  const params = new URLSearchParams({
    q:     query,
    type:  'track',
    limit: String(limit),
  });

  const res = await fetch(`https://api.spotify.com/v1/search?${params}`, {
    headers: { Authorization: `Bearer ${token}` },
  });

  if (res.status === 401) {
    throw new Error('Spotify token expired or invalid — please log in again.');
  }
  if (res.status === 429) {
    const retryAfter = res.headers.get('Retry-After') ?? '?';
    throw new Error(`Spotify rate-limited. Retry after ${retryAfter}s.`);
  }
  if (!res.ok) {
    throw new Error(`Spotify search error ${res.status}: ${await res.text()}`);
  }

  const data = await res.json();
  return data?.tracks?.items ?? [];
}

File: backend/src/resolution/spotify_resolver.py
"""
spotify_resolver.py — Phase 4: Spotify Search

SpotifyResolver.search(title, artist?) → list[Candidate dict]

Three-rung query ladder:
  1. field-qualified  — track:"title" artist:"artist"
  2. plain-combined   — "title" "artist"
  3. title-only       — "title"

Stops at the first rung that returns ≥ 1 result.
Uses the user's OAuth access token (passed in per-request).
"""

from __future__ import annotations
import logging
import re
import httpx

logger = logging.getLogger(__name__)


class SpotifyRateLimitError(Exception):
    """
    Raised when Spotify returns 429 Too Many Requests.

    Carries `retry_after` (seconds, from the Retry-After header) so callers
    can back off correctly instead of treating this as a hard failure —
    a 429 is an expected, recoverable condition, not an error.
    """

    def __init__(self, retry_after: float):
        self.retry_after = retry_after
        super().__init__(f"Spotify rate-limited — retry after {retry_after}s")


# ── Keyword sets for flag derivation (mirrors spotifyResolver.js) ─────────────

LIVE_KEYWORDS = [
    "live", "in concert", "live at", "live from", "live session",
    "live version", "live performance", "acoustic live",
]

REMIX_KEYWORDS = [
    "remix", "remixed", "sped up", "slowed", "nightcore", "flip",
    "bootleg", "edit", "vip", "extended mix", "club mix", "radio edit",
    "mashup", "version", "re-edit",
]


def _contains_keyword(name: str, keywords: list[str]) -> bool:
    lower = name.lower()
    return any(kw in lower for kw in keywords)


def _normalise_track(track: dict, query_rung: str) -> dict:
    """Convert a raw Spotify track object to a Candidate dict."""
    artists = track.get("artists") or []
    images  = (track.get("album") or {}).get("images") or []

    # Prefer the smallest thumbnail (index 2), fall back to larger ones.
    image_url = None
    for idx in (2, 1, 0):
        if idx < len(images):
            image_url = images[idx].get("url")
            break

    release_date = (track.get("album") or {}).get("release_date", "")

    return {
        "id":          track["id"],
        "title":       track["name"],
        "artist":      artists[0]["name"] if artists else None,
        "artists":     ", ".join(a["name"] for a in artists) if artists else None,
        "album":       (track.get("album") or {}).get("name"),
        "imageUrl":    image_url,
        "releaseYear": release_date.split("-")[0] if release_date else None,
        "popularity":  track.get("popularity", 0),
        "durationMs":  track.get("duration_ms", 0),
        "isLive":      _contains_keyword(track["name"], LIVE_KEYWORDS),
        "isRemix":     _contains_keyword(track["name"], REMIX_KEYWORDS),
        "queryRung":   query_rung,
    }


def _build_rungs(title: str, artist: str | None) -> list[dict]:
    """Build the ordered query ladder. Omits artist rungs when artist is None."""
    rungs = []
    if artist:
        rungs.append({
            "label": "field-qualified",
            "query": f'track:"{title}" artist:"{artist}"',
        })
        rungs.append({
            "label": "plain-combined",
            "query": f'"{title}" "{artist}"',
        })
    rungs.append({
        "label": "title-only",
        "query": f'"{title}"',
    })
    return rungs


async def _search_spotify(query: str, limit: int, token: str) -> list[dict]:
    """Raw Spotify /v1/search call. Returns list of track objects."""
    params = {"q": query, "type": "track", "limit": str(limit)}

    async with httpx.AsyncClient(timeout=30.0) as client:
        res = await client.get(
            "https://api.spotify.com/v1/search",
            params=params,
            headers={"Authorization": f"Bearer {token}"},
        )

    if res.status_code == 401:
        raise RuntimeError("Spotify token expired or invalid — please log in again.")
    if res.status_code == 429:
        raw = res.headers.get("Retry-After", "1")
        try:
            retry_after = float(raw)
        except ValueError:
            retry_after = 1.0
        raise SpotifyRateLimitError(retry_after)
    if not res.is_success:
        raise RuntimeError(f"Spotify search error {res.status_code}: {res.text}")

    return res.json().get("tracks", {}).get("items", [])


class SpotifyResolver:
    def __init__(self, limit: int = 8):
        self.limit = limit

    async def search(
        self,
        title: str,
        artist: str | None,
        token: str,
    ) -> list[dict]:
        """
        Search Spotify via the query ladder, return normalised Candidate dicts.

        token — user's OAuth access token (passed through from the request header).
        """
        rungs = _build_rungs(title, artist)

        for rung in rungs:
            tracks = await _search_spotify(rung["query"], self.limit, token)
            if tracks:
                logger.info(
                    '[SpotifyResolver] "%s"%s → rung: %s (%d results)',
                    title,
                    f' / "{artist}"' if artist else "",
                    rung["label"],
                    len(tracks),
                )
                return [_normalise_track(t, rung["label"]) for t in tracks]

            logger.info(
                '[SpotifyResolver] "%s" rung "%s" → 0 results, trying next…',
                title,
                rung["label"],
            )

        logger.warning('[SpotifyResolver] "%s" — all rungs exhausted.', title)
        return []